### ISSUE TITLE
Compatibility: OmniFaces to run on servlet (not Jakarta EE) container

### ISSUE THREAD

**Issue by lprimak** [omnifaces/omnifaces#703]
Hi,

I was trying to run a simple servlet app on Jetty w/OmniFaces on classpath but not even used.
The issue is that OmniFaces has servlet context listener that checks for presence of CDI and JSF. Currently, OmniFaces has great utility classes that don't need CDI or JSF, and I'd like to use them.
However, it's not possible since the context listener breaks deployment.

I would like to change this behavior by printing a warning but still letting the application deploy on a servlet container.

Thank you!

**Comment by BalusC** [comment]
I'd rather not change this default behavior. OmniFaces was and is for Faces projects.

Alternative option: split an OmniServlets or so off OmniFaces and include it in OmniFaces as uberjar -- but if we go down this path, it's probably more correct to *also* split off an OmniUtils (which actually already exist as separate project and shares some existing utils).

Yet another alternative option: introduce context param to skip deployment exception.

**Comment by lprimak** [comment]
Can we start with the context parameter?  Seems easiest. 

Also if we go down the road of splitting the library, I don’t think there would be a need for an Uber jar but just a maven transitive dependency should be good enough.

**Comment by BalusC** [comment]
There exist users who hate dependency managment systems and rather carry around loose JARs :P 

Moreover, the original thought behind OmniFaces is:

> An important design goal will be to have as few dependencies as possible

(from the homepage)

**Comment by lprimak** [comment]
ok :)

**Comment by BalusC** [comment]
Please try the 3.14-SNAPSHOT and let me know. 

The context parameter name is `org.omnifaces.SKIP_DEPLOYMENT_EXCEPTION`. I saw that you already prepped a PR but I don't fully agree. I've implemented the least intrusive approach.

**Comment by lprimak** [comment]
Unfortunately this fails on Jetty while loading the JSP
I think it has to something with OmniFaces still trying to initialize eager beans and faces views.
Here is my test: https://github.com/flowlogix/shiro-smoketest

- Change pom.xml omnifaces version from `3.13.4-fl-2` to `3.14-SNAPSHOT`
- enable snapshots
`mvn -Pjetty`
```
URI: | /auth/login.jsp
500
javax.servlet.ServletException: org.apache.jasper.JasperException: javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
jsp
javax.servlet.ServletException: org.apache.jasper.JasperException: javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
org.apache.jasper.JasperException: javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
java.lang.IllegalStateException: Create session failed

<h3 style="font-style: normal; font-variant-caps: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0);">Caused by:</h3><pre style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0);">javax.servlet.ServletException: org.apache.jasper.JasperException: javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:162)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:516)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.apache.jasper.JasperException: javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
	at org.apache.jasper.servlet.JspServletWrapper.handleJspException(JspServletWrapper.java:618)
	at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:488)
	at org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:383)
	at org.apache.jasper.servlet.JspServlet.service(JspServlet.java:331)
	at org.eclipse.jetty.jsp.JettyJspServlet.service(JettyJspServlet.java:106)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1450)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
	at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1656)
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:292)
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
	at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:61)
	at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
	at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
	at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
	at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
	at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
	at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
	at org.apache.shiro.web.servlet.AbstractShiroFilter.executeChain(AbstractShiroFilter.java:450)
	at com.flowlogix.shiro.ee.filters.ShiroFilter.executeChain(ShiroFilter.java:238)
	at org.apache.shiro.web.servlet.AbstractShiroFilter$1.call(AbstractShiroFilter.java:365)
	at org.apache.shiro.subject.support.SubjectCallable.doCall(SubjectCallable.java:90)
	at org.apache.shiro.subject.support.SubjectCallable.call(SubjectCallable.java:83)
	at org.apache.shiro.subject.support.DelegatingSubject.execute(DelegatingSubject.java:387)
	at org.apache.shiro.web.servlet.AbstractShiroFilter.doFilterInternal(AbstractShiroFilter.java:362)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
	... 12 more
Caused by: javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
	at org.apache.jsp.auth.login_jsp._jspService(login_jsp.java:158)
	at org.apache.jasper.runtime.HttpJspBase.service(HttpJspBase.java:70)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:465)
	... 57 more
Caused by: java.lang.IllegalStateException: Create session failed
	at org.eclipse.jetty.server.Request.getSession(Request.java:1630)
	at javax.servlet.http.HttpServletRequestWrapper.getSession(HttpServletRequestWrapper.java:270)
	at org.apache.shiro.web.servlet.ShiroHttpServletRequest.getSession(ShiroHttpServletRequest.java:148)
	at org.apache.shiro.web.servlet.ShiroHttpServletRequest.getSession(ShiroHttpServletRequest.java:190)
	at org.apache.jasper.runtime.PageContextImpl.initialize(PageContextImpl.java:134)
	at org.apache.jasper.runtime.JspFactoryImpl.internalGetPageContext(JspFactoryImpl.java:109)
	at org.apache.jasper.runtime.JspFactoryImpl.getPageContext(JspFactoryImpl.java:59)
	at org.apache.jsp.auth.login_jsp._jspService(login_jsp.java:115)
	... 60 more
</pre><hr style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0);"><a href="https://eclipse.org/jetty" style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px;">Powered by Jetty:// 9.4.48.v20220622</a><br class="Apple-interchange-newline" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;">HTTP ERROR 500 javax.servlet.ServletException: org.apache.jasper.JasperException: javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed

URI:	/auth/login.jsp
STATUS:	500
MESSAGE:	javax.servlet.ServletException: org.apache.jasper.JasperException: javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
SERVLET:	jsp
CAUSED BY:	javax.servlet.ServletException: org.apache.jasper.JasperException: javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
CAUSED BY:	org.apache.jasper.JasperException: javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
CAUSED BY:	javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
CAUSED BY:	java.lang.IllegalStateException: Create session failed
Caused by:

javax.servlet.ServletException: org.apache.jasper.JasperException: javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:162)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:516)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.apache.jasper.JasperException: javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
	at org.apache.jasper.servlet.JspServletWrapper.handleJspException(JspServletWrapper.java:618)
	at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:488)
	at org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:383)
	at org.apache.jasper.servlet.JspServlet.service(JspServlet.java:331)
	at org.eclipse.jetty.jsp.JettyJspServlet.service(JettyJspServlet.java:106)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1450)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
	at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1656)
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:292)
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
	at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:61)
	at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
	at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
	at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
	at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
	at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
	at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
	at org.apache.shiro.web.servlet.AbstractShiroFilter.executeChain(AbstractShiroFilter.java:450)
	at com.flowlogix.shiro.ee.filters.ShiroFilter.executeChain(ShiroFilter.java:238)
	at org.apache.shiro.web.servlet.AbstractShiroFilter$1.call(AbstractShiroFilter.java:365)
	at org.apache.shiro.subject.support.SubjectCallable.doCall(SubjectCallable.java:90)
	at org.apache.shiro.subject.support.SubjectCallable.call(SubjectCallable.java:83)
	at org.apache.shiro.subject.support.DelegatingSubject.execute(DelegatingSubject.java:387)
	at org.apache.shiro.web.servlet.AbstractShiroFilter.doFilterInternal(AbstractShiroFilter.java:362)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
	... 12 more
Caused by: javax.servlet.ServletException: java.lang.IllegalStateException: Create session failed
	at org.apache.jsp.auth.login_jsp._jspService(login_jsp.java:158)
	at org.apache.jasper.runtime.HttpJspBase.service(HttpJspBase.java:70)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:465)
	... 57 more
Caused by: java.lang.IllegalStateException: Create session failed
	at org.eclipse.jetty.server.Request.getSession(Request.java:1630)
	at javax.servlet.http.HttpServletRequestWrapper.getSession(HttpServletRequestWrapper.java:270)
	at org.apache.shiro.web.servlet.ShiroHttpServletRequest.getSession(ShiroHttpServletRequest.java:148)
	at org.apache.shiro.web.servlet.ShiroHttpServletRequest.getSession(ShiroHttpServletRequest.java:190)
	at org.apache.jasper.runtime.PageContextImpl.initialize(PageContextImpl.java:134)
	at org.apache.jasper.runtime.JspFactoryImpl.internalGetPageContext(JspFactoryImpl.java:109)
	at org.apache.jasper.runtime.JspFactoryImpl.getPageContext(JspFactoryImpl.java:59)
	at org.apache.jsp.auth.login_jsp._jspService(login_jsp.java:115)
	... 60 more
[Powered by Jetty:// 9.4.48.v20220622](https://eclipse.org/jetty)
```

**Comment by lprimak** [comment]
Also, OmniFaces should not issue any warnings if the flag is set, because the user clearly doesn't want anything initialized at that point.
Basically, as it was done by me originally :)

**Comment by lprimak** [comment]
Basically, as it was done by me originally :)

**Comment by BalusC** [comment]
> because the user clearly doesn't want anything initialized at that point.

This conflicts the fact that OmniFaces is an utility library for JSF/Faces.

I've removed the stack trace from the log and swapped EagerBeans/FacesViews. 

I get only this on an empty Tomcat and the rest works fine for me:
```
Sep 17, 2022 3:58:51 PM org.omnifaces.ApplicationInitializer onStartup
WARNING: OmniFaces failed to initialize! java.lang.NoClassDefFoundError: javax/faces/webapp/FacesServlet
Sep 17, 2022 3:58:51 PM org.omnifaces.ApplicationListener contextInitialized
WARNING: OmniFaces failed to initialize! java.lang.NoClassDefFoundError: javax/faces/application/ViewHandler
```

**Comment by lprimak** [comment]
I'll double check w/the latest

**Comment by lprimak** [comment]
Just tried it with Jetty. Unfortunately the same error persists.

**Comment by melloware** [comment]
@lprimak Here is a reproducer you can run `mvn clean jetty:run` with.  I am not seeing the warnings above?
[O703.zip](https://github.com/omnifaces/omnifaces/files/9593851/O703.zip)

**Comment by BalusC** [comment]
@melloware: you need to remove all JSF and CDI things from `web.xml`. Trim down `web.xml` until you have only the `org.omnifaces.SKIP_DEPLOYMENT_EXCEPTION` context param. Also the CDI bean manager needs to be removed from `jetty-env.xml`. This gives me the following log:
```
[INFO] --- jetty-maven-plugin:9.4.48.v20220622:run (default-cli) @ primefaces-test ---
[INFO] Logging initialized @1534ms to org.eclipse.jetty.util.log.Slf4jLog
[INFO] webAppSourceDirectory not set. Trying src/main/webapp
[INFO] Reload Mechanic: automatic
[INFO] nonBlocking:false
[INFO] Classes = /home/bauke/Downloads/O703/target/classes
[INFO] Configuring Jetty for project: primefaces-test
[INFO] Context path = /primefaces-test
[INFO] Tmp directory = /home/bauke/Downloads/O703/target/tmp
[INFO] Web defaults = org/eclipse/jetty/webapp/webdefault.xml
[INFO] Web overrides =  none
[INFO] web.xml file = file:///home/bauke/Downloads/O703/target/web.xml
[INFO] Webapp directory = /home/bauke/Downloads/O703/src/main/webapp
[INFO] jetty-9.4.48.v20220622; built: 2022-06-21T20:42:25.880Z; git: 6b67c5719d1f4371b33655ff2d047d24e171e49a; jvm 11.0.12+7
[INFO] Scanning elapsed time=81ms
Sep 18, 2022 8:39:46 AM org.omnifaces.ApplicationInitializer logOmniFacesVersion
INFO: Using OmniFaces version 3.14-SNAPSHOT
Sep 18, 2022 8:39:46 AM org.omnifaces.ApplicationInitializer onStartup
WARNING: OmniFaces failed to initialize! java.lang.NoClassDefFoundError: javax/faces/webapp/FacesServlet
[INFO] DefaultSessionIdManager workerName=node0
[INFO] No SessionScavenger set, using defaults
[INFO] node0 Scavenging every 600000ms
Sep 18, 2022 8:39:46 AM org.omnifaces.ApplicationListener contextInitialized
WARNING: OmniFaces failed to initialize! java.lang.NoClassDefFoundError: javax/faces/application/ViewHandler
[INFO] Started o.e.j.m.p.JettyWebAppContext@234cff57{/primefaces-test,[file:///home/bauke/Downloads/O703/src/main/webapp/, jar:file:///home/bauke/.m2/repository/org/omnifaces/omnifaces/3.14-SNAPSHOT/omnifaces-3.14-SNAPSHOT.jar!/META-INF/resources],AVAILABLE}{file:///home/bauke/Downloads/O703/src/main/webapp/}
[INFO] Started ServerConnector@4b765e92{HTTP/1.1, (http/1.1)}{0.0.0.0:8080}
[INFO] Started @2083ms
[INFO] Started Jetty Server
```

**Comment by lprimak** [comment]
@BalusC can you reproduce the exceptions [above](https://github.com/omnifaces/omnifaces/issues/703#issuecomment-1242794220) I am getting with Jetty? It happens when loading any JSP

**Comment by BalusC** [comment]
Nope, no exceptions with below JSP
```
<!DOCTYPE html>
<html>
    <head>
        <title>Test</title>
    </head>
    <body>
        <h1>Test</h1>
        <p>${pageContext.session.id}</p>
    </body>
</html>
```

**Comment by lprimak** [comment]
Can you attach your reproduced please?

**Comment by lprimak** [comment]
Can you please try with my reproducer? I stripped everything out besides Omnifaces and it's still failing.
Or attach pom.xml?

**Comment by lprimak** [comment]
Ok, I now understand what's happening.
I had `jakarta-web-api` on Jetty classpath which prevented FacesViews initialization from failing, and thus would continue to call eager beans which (in a very non-obvious way) causes the errors.

Here's the relevant pom.xml excerpt (removes provided scope and makes it compile-scope):
```
    <dependencies>
        <dependency>
            <groupId>javax</groupId>
            <artifactId>javaee-web-api</artifactId>
            <scope>provided</scope>
        </dependency>
     </dependencies>
        <profile>
            <id>jetty</id>
            <build>
                <defaultGoal>jetty:run</defaultGoal>
            </build>
            <dependencies>
                <dependency>
                    <groupId>javax</groupId>
                    <artifactId>javaee-web-api</artifactId>
                </dependency>
                <dependency>
                    <groupId>jakarta.enterprise</groupId>
                    <artifactId>jakarta.enterprise.cdi-api</artifactId>
                </dependency>
            </dependencies>
        </profile>
    </profiles>
```

This solution does feel brittle, i.e. relying on one initialization to fail to not do others (which fail in a worse way) like Eager beans. I would still prefer that the initialization is not called at all if the context parameter is present.
However, it's no longer a showstopper, even though it's brittle.

You can close this issue or implement my suggestion (which I would prefer)
thanks!

**Comment by BalusC** [comment]
Jetty does not provide Java EE web API. It only provides a small handful parts thereof (Servlet API, JSP API, EL API, WebSocket API and probably Security API, don't know off top of head).

I'll try reproducing this specific case caused by a misconfigured pom.xml and look what I can do wrt Eager beans initialization.

For your information, the `provided` dependencies in `pom.xml` of my test project targeted at [Tomcat 9](https://tomcat.apache.org/whichversion.html) server looks like this:
```
		<dependency>
			<groupId>javax.servlet</groupId>
			<artifactId>javax.servlet-api</artifactId>
			<version>4.0.0</version>
			<scope>provided</scope>
		</dependency>
		<dependency>
			<groupId>javax.servlet.jsp</groupId>
			<artifactId>javax.servlet.jsp-api</artifactId>
			<version>2.3.0</version>
			<scope>provided</scope>
		</dependency>
		<dependency>
			<groupId>javax.el</groupId>
			<artifactId>javax.el-api</artifactId>
			<version>3.0.0</version>
			<scope>provided</scope>
		</dependency>
		<dependency>
			<groupId>javax.websocket</groupId>
			<artifactId>javax.websocket-api</artifactId>
			<version>1.1</version>
			<scope>provided</scope>
		</dependency>
		<dependency>
			<groupId>javax.security.enterprise</groupId>
			<artifactId>javax.security.enterprise-api</artifactId>
			<version>1.0</version>
			<scope>provided</scope>
		</dependency>

```

**Comment by lprimak** [comment]
Thanks!  
just to make my use case clearer, the intent is to use Jakarta EE features when the server supports it or to fall back to a subset of features when running on Jetty or other servlet container

**Comment by BalusC** [comment]
Makes sense.

**Comment by BalusC** [comment]
I reproduced the issues with Eager Beans initialization and JSP session creation only when I added Java EE API as `compile` time dependency. This is just wrong. It should never be marked as `compile` time dependency unless you're developing/implementing the whole Java EE server itself (e.g. WildFly). For web applications it should _always_ be marked `provided`. This worked fine for me. I did notice that Jetty spitted a large list of warnings that certain Java EE artifacts were found in multiple locations (in Jetty itself and in the packaged Java EE dependency). That should already have given away a clue that dependency config was wrong.

I have improved/rearranged the initialization a bit via  https://github.com/omnifaces/omnifaces/commit/e4edef106f406a977d4e56cffbdc268a5157bb50. Now it merely prints the following in Jetty when you package the whole Java EE API in the webapp itself instead of letting the server provide it:
```
Sep 19, 2022 2:07:04 PM org.omnifaces.ApplicationListener contextInitialized
WARNING: OmniFaces failed to initialize! java.lang.IllegalStateException: CDI BeanManager instance is not available in this environment.
```

The technical problem with JSP session creation was because Eager Beans forcibly installed a `HttpSessionListener` with the expectation that CDI impl will possibly be available later on, which in the end didn't work at all. That `HttpSessionListener` basically threw the same NPE again as during Eager Beans initialization fail, but this was in turn completely suppressed from the exception/stacktrace/logs during JSP session creation.
